tao_compiler/mlir/disc/transforms/fusion_utils.h

```diff
@@ -227,88 +236,124 @@ class FusionPattern {
     return value.getDefiningOp();
   }
 
- private:
-  FusionPattern(SmallVectorImpl<Operation*>& op_list);
+  void updateLastWriter(Value value, Operation* op) {
+    last_writer_[value] = op;
+  }
+
+ protected:
+  FusionPatternBase(SmallVectorImpl<Operation*>& op_list);
 
- private:
   // Calculates the inputs and outputs of the fusion pattern.
   void calculateOperandsAndResults();
 
- private:
   FusionOpList op_list_;
-  Operation* dominant_op_ = nullptr;
-  FusionType fusion_type_ = FusionType::kNone;
   FusionValueList operands_;
   FusionValueList results_;
   FusionValueList internal_results_;
+  FusionValueList external_only_results_;
   SmallVector<Operation*, 4> root_ops_;
   DenseMap<Value, Operation*> last_writer_;
 };
 
-// Represents a list of disjoint fusion patterns for a block.
-using FusionPlan = std::vector<FusionPattern>;
+// Represents a list of lmhlo ops that are going to be fused.
+class FusionPattern : public FusionPatternBase {
+ public:
+  // Create a new fusion pattern from a single op.
+  explicit FusionPattern(Operation* op);
 
-using llvm::EquivalenceClasses;
+  // Create a new fusion pattern from the ops inside the lmhlo fusion op.
+  explicit FusionPattern(lmhlo::FusionOp op, ShapeAnalysis* shape_analysis);
 
-// Supports using EquivalenceClasses for Value
-class ValueWrapper {
- public:
-  explicit ValueWrapper(Value value) : value_(std::move(value)) {}
+  // Do not allow to build a fusion pattern with only FusionOp.
+  explicit FusionPattern(lmhlo::FusionOp op) = delete;
```

Comment:
The above method `explicit FusionPattern(lmhlo::FusionOp op, ShapeAnalysis* shape_analysis);` does not has default parameter value, so the deleted version is not actually needed?

Reply:
To check.

Reply:
There is FusionPatternBase(lmhlo::FusionOp op)